app/factory/Layer.js

```diff
@@ -914,87 +914,110 @@ Ext.define('CpsiMapview.factory.Layer', {
-                    origTreeNodeConf = CpsiMapview.controller.LayerTreeController.getTreeNodeConf(newLayer.get('layerKey'));
-                    newLayer.set('_origTreeConf', origTreeNodeConf);
+                if(isSwitchLayer){
+                    switchNecessary = LayerFactory.isLayerSwitchNecessary(layer, resolution);
                 }
 
-                if (newLayer.get('isWms')) {
+                if(isSwitchLayer && switchNecessary){
 
-                    activeStyle = LegendUtil.getWmsStyleFromSldFile(activeStyle);
+                    var switchConfiguration = layer.get('switchConfiguration');
+                    // restore current layer visibility
+                    switchConfiguration.visibility = layer.getVisible();
+                    // also apply current filter and selected style
 
-                    // check if a label STYLES parameter was added --> keep this
-                    // the STYLES value (SLD) for the labels
-                    var labelClassName = newLayer.get('labelClassName');
-                    if (newLayer.get('labelsActive') === true) {
-                        activeStyle += ',' + labelClassName;
-                    }
+                    var activeStyle = layer.get('activatedStyle');
+                    var filters = layer.getSource().get('additionalFilters');
+
+                    var newLayer = LayerFactory.createSwitchLayer(switchConfiguration);
+                    var newLayerSource = newLayer.getSource();
+                    // store filters for either layer type so they can be retrieved when switching
+                    newLayerSource.set('additionalFilters', filters);
 
-                    if (filters && filters.length > 0) {
-                        newLayerSource.getParams().FILTER = GeoExt.util.OGCFilter.getOgcFilterFromExtJsFilter(filters, 'wms', 'and', '1.1.0');
+                    // add original tree config (from tree.json) to new layer
+                    var origTreeNodeConf = newLayer.get('_origTreeConf');
+                    if (!origTreeNodeConf) {
+                        origTreeNodeConf = CpsiMapview.controller.LayerTreeController.getTreeNodeConf(newLayer.get('layerKey'));
+                        newLayer.set('_origTreeConf', origTreeNodeConf);
                     }
 
-                    // ensure there is a filter for every layer listed in the WMS request (required by MapServer)
-                    var wmsFilterUtil = CpsiMapview.util.WmsFilter;
-                    var wmsFilterString = wmsFilterUtil.getWmsFilterString(newLayer);
+                    if (newLayer.get('isWms')) {
 
-                    // apply new style parameter and reload layer
-                    var newParams = {
-                        STYLES: activeStyle,
-                        FILTER: wmsFilterString,
-                        cacheBuster: Math.random()
-                    };
+                        activeStyle = LegendUtil.getWmsStyleFromSldFile(activeStyle);
```

Comment:
Error thrown here as `activeStyle` is undefined. This is when the application starts zoomed in, and the switch layer is off by default, and then zoomed out so the WMS takes over. This happens if all switch layers are switched off and the map is simply zoomed in and out.